### PR TITLE
8256517: (ref) Reference.clear during reference processing may lose notification

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change to Reference.clear() to address several issues.

(JDK-8240696) For GCs using a SATB barrier, simply assigning the referent
field to null may extend the lifetime of the referent value.

(JDK-8240696) For GCs with concurrent reference processing, clearing the
referent field during reference processing may discard the expected
notification.

Both of these are addressed by introducing a private native helper function
for clearing the referent, rather than using an ordinary in-Java field
assignment. Tests have been added for both of these issues. This required
adding a new breakpoint in reference processing for ZGC.

Of course, finalization adds some complexity to the problem.  We deal with
that by having FinalReference override clear.  The implementation is
provided by a new package-private method in Reference.  (There are a number
of alternatives, all of them clumsy; finalization is annoying that way.)

While dealing with FinalReference clearing it was noted that the recent
JDK-8256106 and JDK-8256370 have some problems. FinalizerHistogram was not
updated to call the new Reference.getInactive(), instead still calling get()
on FinalReferences, with the JDK-8256106 problems. Fixing that showed the
assertion for inactive FinalReference added by JDK-8256370 used the wrong
test.

Rather than tracking down and changing all get() and clear() calls on final
references and changing them to use getInactive and a new similar clear
function, I've changed FinalReference to override get and clear, which call
the helper functions in Reference. I've also renamed getInactive to be more
explanatory and less convenient to call directly, and similarly named the
helper for clear. This means that get/clear should never be called on an
active FinalReference. That's already never done, and would have problems
if it were.

Testing:
mach5 tier1-6
Local (linux-x64) tier1 using Shenandoah.
New TestReferenceClearDuringMarking fails for G1 without these changes.
New TestReferenceClearDuringReferenceProcessing fails for ZGC without these changes.

/issue JDK-8256517, JDK-8240696
/label core-libs, hotspot
/summary Use private native helper to implement Reference.clear.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issues
 * [JDK-8256517](https://bugs.openjdk.java.net/browse/JDK-8256517): (ref) Reference.clear during reference processing may lose notification
 * [JDK-8240696](https://bugs.openjdk.java.net/browse/JDK-8240696): (ref) Reference.clear may extend the lifetime of the referent

### Reviewers
 * [Per Liden](https://openjdk.java.net/census#pliden) (@pliden - **Reviewer**) ⚠️ Review applies to c19efd7006d89b8cc18116076453517474332e41
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to 46e5b1f6d1a04aaf630aac58f58051f198b21b0a
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to 46e5b1f6d1a04aaf630aac58f58051f198b21b0a

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1376/head:pull/1376`
`$ git checkout pull/1376`

### CODE REVIEW

**Review by pliden — CHANGES_REQUESTED** [review]
Looks good. Just want to request that you also remove the following comment in zReferenceProcessor.cpp, as it's no longer true.

```--- a/src/hotspot/share/gc/z/zReferenceProcessor.cpp
+++ b/src/hotspot/share/gc/z/zReferenceProcessor.cpp
@@ -184,12 +184,6 @@ bool ZReferenceProcessor::should_discover(oop reference, ReferenceType type) con
 }
 
 bool ZReferenceProcessor::should_drop(oop reference, ReferenceType type) const {
-  // This check is racing with a call to Reference.clear() from the application.
-  // If the application clears the reference after this check it will still end
-  // up on the pending list, and there's nothing we can do about that without
-  // changing the Reference.clear() API. This check is also racing with a call
-  // to Reference.enqueue() from the application, which is unproblematic, since
-  // the application wants the reference to be enqueued anyway.
   const oop referent = reference_referent(reference);
   if (referent == NULL) {
     // Reference has been cleared, by a call to Reference.enqueue()
```

File: test/hotspot/jtreg/gc/TestReferenceClearDuringReferenceProcessing.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package gc;
+
+/* @test
+ * @bug 8256517
+ * @requires vm.gc.Z
```
Comment (rkennke) [inline test/hotspot/jtreg/gc/TestReferenceClearDuringReferenceProcessing.java:28]: Please add | vm.gc.Shenandoah here.

**Review by rkennke — CHANGES_REQUESTED** [review]
Looks good, except one small nit in one of the test configs.

File: test/hotspot/jtreg/gc/TestReferenceClearDuringReferenceProcessing.java (hunk 1 above)
Comment (pliden) [reply test/hotspot/jtreg/gc/TestReferenceClearDuringReferenceProcessing.java:28]: Note that for this test to be useful, the GC needs to support concurrent GC breakpoints, which Shenandoah doesn't do.

Comment (rkennke) [reply test/hotspot/jtreg/gc/TestReferenceClearDuringReferenceProcessing.java:28]: Ok, right. Nevermind then!

**Review by rkennke — APPROVED** [review]
Looks good! Thanks!

**Review by mlchung — APPROVED** [review]
Looks good.  Thanks for correcting `assert next != null`  inactive FinalReference check.

**Comment by kimbarrett** [comment]
> Looks good. Just want to request that you also remove the following comment in zReferenceProcessor.cpp, as it's no longer true.
> 
> ```
> --- a/src/hotspot/share/gc/z/zReferenceProcessor.cpp
> +++ b/src/hotspot/share/gc/z/zReferenceProcessor.cpp
> @@ -184,12 +184,6 @@ bool ZReferenceProcessor::should_discover(oop reference, ReferenceType type) con
>  }
>  
>  bool ZReferenceProcessor::should_drop(oop reference, ReferenceType type) const {
> -  // This check is racing with a call to Reference.clear() from the application.
> -  // If the application clears the reference after this check it will still end
> -  // up on the pending list, and there's nothing we can do about that without
> -  // changing the Reference.clear() API. This check is also racing with a call
> -  // to Reference.enqueue() from the application, which is unproblematic, since
> -  // the application wants the reference to be enqueued anyway.
>    const oop referent = reference_referent(reference);
>    if (referent == NULL) {
>      // Reference has been cleared, by a call to Reference.enqueue()
> ```

Done.

**Comment by kimbarrett** [comment]
I realized there was a theoretical problem with the new tests; they
should also be ensuring the Reference objects are in oldgen.  That's
fixed in the latest push.

**Review by pliden — APPROVED** [review]
Looks good!

**Comment by kimbarrett** [comment]
Thanks @rkennke, @pliden, @mlchung for reviews.

**Pull request closed without merging**